[webkitbugspy] Implement ability to relate issues

briannafan opened the pull request:
#### b88a46556b32c790b964a7fd4989979586c3b216
<pre>
[webkitbugspy] Implement ability to relate issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=266432">https://bugs.webkit.org/show_bug.cgi?id=266432</a>
<a href="https://rdar.apple.com/problem/119679931">rdar://problem/119679931</a>

Reviewed by Jonathan Bedard.

Relations from bugs on bugzilla can now be retrieved and set. Radars coming!

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate):
(Tracker.relate):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.__init__):
(Issue):
(Issue.related):
(Issue.relate):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla._issue):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker.relate):

Canonical link: <a href="https://commits.webkit.org/272788@main">https://commits.webkit.org/272788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3a066c583e827f811275d4ae6ac39c5c9db5595

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33072 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11849 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34981 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35713 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/29874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/34042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/14189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/9016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33547 "Failed to checkout and rebase branch from PR 21899") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/9979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/29541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/8703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/33347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/29508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/37044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/30042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/29908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/37044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/8995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/6934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/37044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/10695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4253 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/9635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -248,6 +248,13 @@ def populate(self, issue, member=None):
                 issue._version = response.get('version', '')
                 issue._keywords = response.get('keywords', [])
 
+                related_dict = dict()
```
Should probably construct this directly instead of line-by-line:
```
issue._related = dict(
    blocks=response.get('blocks') or [],
    ...
)
```

Another question I have here: what is the type of `issue._related` supposed to be? I would expect it to be issue objects, but that's clearly not what `response.get` is supposed to be. Note that because issues are lazily populated, assuming `blocks` is a list of Bugzilla IDs, you can safely construct `Issue` objects without worrying about network requests until code starts using the issue.

briannafan replied:
It's a dictionary where keys are the relationship and values are a list of integer bug ids. 
`issue._related = {'depends_on': [], 'blocks': [], 'regressions': [], 'regressed_by': []}`
Do you mean that we can add `self.issue(bug_id)` to the lists instead of just `bug_id`? Now that I'm thinking about it, this might be a better place to construct `Issue`s than whenever an engineer might be trying to use `issue._related`.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -461,6 +468,49 @@ def add_comment(self, issue, text):
+
+        update_dict = dict()
+        update_dict['ids'] = [issue.id]
+        if depends_on:
```
This should probably be an `elif`, right? Also, more generally, it's not correct to try and define multiple relationships at a time, is it? If it is, then this function should really accept lists of issues, if it's not, then we should have code that validates the user only passed a single issue. That probably points to accepting `**relations` and then asserting there is only a single key, then validating the key is one of the expected inputs.

briannafan replied:
The API supports setting multiple relationships in one request, so I wanted to add that as an option. issue should be a singular issue, whereas the keyword args are all supposed to be lists of issues. Should I add type hints to make that clear?

JonWBedard replied:
We still have Python 2 support here, so we have to be careful with type hints.

Because the API supports setting multiple, this is reasonable, although it is probably better to specifically iterate over the arguments and extract IDs rather than expecting the caller to pass integer IDs instead of `Issue` objects. Might also want to check that the `Issue` objects in question are actually from this tracker as well....

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -461,6 +468,49 @@ def add_comment(self, issue, text):
+                json=update_dict,
+                timeout=self.timeout,
+            )
+        except RuntimeError as e:
```
Does `self.session.puts`'s exception inherit from `RuntimeError`?

briannafan replied:
To be honest, I took the error handling from existing code in `set`. 
`self.session.puts` inherits from `requests.exceptions.RequestException` and those errors are handled in the next two conditional statements. Should that be changed?

JonWBedard replied:
We should catch `requests.exceptions.RequestException` specifically, I think. The issue here isn't the normal exceptions (which generally don't generate exceptions unless you ask them to), but more exotic things like network timeouts or sockets closing on you.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -461,6 +468,49 @@ def add_comment(self, issue, text):
+            sys.stderr.write("Failed to modify '{}'\n".format(issue))
+            return None
+
+        if not issue._related:
```
Should probably have an `else` here, I think. No need to explicitly update `issue._related` if we had to populate it, right?

briannafan replied:
Good catch

JonWBedard reviewed (commented):
I think this is quite close! I think the two big things that you need to decide:
- What does `related` return? It seems like you intend it to be a dictionary of relation types. That probably means a dictionary where keys are relationships, follows by a list of issue objects with that relationship. That's not the way you've implemented this so far.
- How are the relationship types going to translate between issue trackers? This is probably more important for writing than it is for reading.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -461,6 +466,66 @@ def add_comment(self, issue, text):
         return result
 
+    def issues_to_ids(self, issues):
+        if not isinstance(issues, list):
```
It could be a `tuple` too, right?

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -461,6 +466,66 @@ def add_comment(self, issue, text):
+        issue_ids = []
+        for i in issues:
+            try:
+                issue_id = i.link
```
Not quite sure what the purpose of this check is? Since we actually return the `issue_ids` list, right?

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -461,6 +466,66 @@ def add_comment(self, issue, text):
+                issue_id = i.link
+            except AttributeError:
+                raise AttributeError('Input should be a list of Issue objects.')
+            if isinstance(i.tracker, Tracker):
```
This check should be sufficient, because every issue which has a Bugzilla Tracker will have an ID. And if the caller provides something other than an `issue` object, you'll end up with an `AttributeError` anyways. All of this is to say, I don't think you need the `i.link` check.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -461,6 +466,66 @@ def add_comment(self, issue, text):
+        except requests.exceptions.RequestException as e:
+            sys.stderr.write('Request Error: {}\n'.format(e))
+        except RuntimeError as e:
+            sys.stderr.write('Runtime Error:{}\n'.format(e))
```
I don't think we should catch a runtime error here, I don't see how we would raise one.

briannafan committed:
[webkitbugspy] Implement ability to relate issues
https://bugs.webkit.org/show_bug.cgi?id=266432
rdar://problem/119679931

Reviewed by Jonathan Bedard.

Relations from bugs on bugzilla can now be retrieved and set. Radars coming!

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate):
(Tracker.relate):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.__init__):
(Issue):
(Issue.related):
(Issue.relate):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla._issue):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker.relate):

Canonical link: https://commits.webkit.org/272788@main
